smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -155,150 +347,314 @@ def _allocate_step(
+                            redir_group = None
+                        except Exception as e:
+                            logger.error(e)
+
+                self._group_infos[step_id].status = SmartSimStatus.STATUS_CANCELLED
+                self._group_infos[step_id].return_codes = [-9]
+
+    def _start_steps(self) -> None:
+        self._heartbeat()
+        with self._queue_lock:
+            started = []
+            for step_id, request in self._queued_steps.items():
+                hosts = self._allocate_step(step_id, self._queued_steps[step_id])
+                if not hosts:
+                    continue
+
+                logger.debug(f"Step id {step_id} allocated on {hosts}")
+
+                global_policy = dragon_policy.Policy(
+                    placement=dragon_policy.Policy.Placement.HOST_NAME,
+                    host_name=hosts[0],
+                )
+                grp = dragon_process_group.ProcessGroup(
+                    restart=False, pmi_enabled=request.pmi_enabled, policy=global_policy
+                )
+
+                policies = []
+                for node_name in hosts:
+                    local_policy = dragon_policy.Policy(
+                        placement=dragon_policy.Policy.Placement.HOST_NAME,
+                        host_name=node_name,
+                    )
+                    policies.extend([local_policy] * request.tasks_per_node)
+                    tmp_proc = dragon_process.ProcessTemplate(
+                        target=request.exe,
+                        args=request.exe_args,
+                        cwd=request.path,
+                        env={**request.current_env, **request.env},
+                        stdout=dragon_process.Popen.PIPE,
+                        stderr=dragon_process.Popen.PIPE,
+                        policy=local_policy,
+                    )
+                    grp.add_process(nproc=request.tasks_per_node, template=tmp_proc)
+
+                try:
+                    grp.init()
+                    grp.start()
+                    grp_status = SmartSimStatus.STATUS_RUNNING
+                except Exception as e:
+                    logger.error(e)
+                    grp_status = SmartSimStatus.STATUS_FAILED
+
+                puids = None
+                try:
+                    puids = list(
+                        set(grp.puids + [puid for puid, retcode in grp.inactive_puids])
+                    )
+                    self._group_infos[step_id] = ProcessGroupInfo(
+                        process_group=grp,
+                        puids=puids,
+                        return_codes=[],
+                        status=grp_status,
+                        hosts=hosts,
+                    )
+                    self._running_steps.append(step_id)
+                    started.append(step_id)
+                except Exception as e:
+                    logger.error(e)
+
+                if (
+                    puids is not None
```

Comment:
- If we end up with `len(puids) != len(policies)` it seems like something is wonky, but we don't report it.